### DISCUSSION TITLE
The Error of Layer Number

### DISCUSSION THREAD

**Discussion by 706697916** [The-OpenROAD-Project/OpenROAD-flow-scripts#1676] [General]
I changed to a new PDK using ORFS. The layer number of M1 to M5 is 101 to105 and the via number of V1 to V4 is 121 to 124. I write it to the klayout tech file(.lyt). But the final gds file of ORFS has wrong number 10 to 18 for M1 to M5 and V1 to V4. The picture below shows the wrong number in the klayout.
If I use the "make gui_final" command, the gui will give me the right layer name of M1 to M5 and V1 to V4.
I wonder where to change the corresponding number of different layers.
![2b694590528c1c47eb76eb23fb9186f](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/1b801d39-d3b2-462f-8703-06862c04e994)
![c6103daaf24a01677f400eaba3c22d0](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/09928101-1273-48b8-a8bc-db4be1fe8503)

**Comment by maliberty** [comment]
Most likely your problem is with your GDS_LAYER_MAP which determines the conversion.  The .lyt is just display settings.

**Reply by 706697916** [reply]
Where to change it?

**Reply by maliberty** [reply]
In whatever file you are pointing the variable GDS_LAYER_MAP  to in the setup for your platform

**Reply by rovinski** [reply]
@maliberty other way around: `.lyp` is display settings, `.lyt` is the technology file.
Also it seems like gf180 is the only one really using `GDS_LAYER_MAP`. The other PDKs just specify it in the `.lyt` file

@706697916 you can look at e.g. nangate45 lyt as an example on how to specify the layer map using a text editor: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/1cdd4e7165c70ea59c06ab05150644f5d09060d1/flow/platforms/nangate45/FreePDK45.lyt#L24

Alternatively you can use the KLayout GUI to modify the layermap as mentioned in the KLayout docs: https://www.klayout.de/doc/about/lef_def_import.html

**Reply by maliberty** [reply]
Thanks @rovinski that's a more complete answer.

**Reply by 706697916** [reply]
If I export the "GDS_LAYER_MAP" in the platform config file, the ORFS will report "no match via gds cell" error. 

And I change to write the layer number in the .lyt file, the ORFS will give me the correct M1, V1, M2 layer number, but it still shows the error of wrong layer number of M3 M4 M5 and V2 V3 V4. It is interesting that M1, V1 and M2 layer have both the correct layer number and a wrong number of 10, 11, 12.

I think the input gds file of my plateform is the reason why V2 M2......V4 M5 layer cannot be correctly identified. My input gds file does not contain these layers. But the input gds file of asap7 also does not contain all of the metal and via layer. I wonder why asap7 platform has an output gds file of correct layer number.
![1](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/27b6f863-4fb7-49e4-9487-1baeb4d924e5)
![2](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/cce35b11-7293-4f2d-93bb-95c2ef1f7285)
![3](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/b4af837e-d2bd-41f7-94ca-117792a78e7c)

**Comment by maliberty** [comment]
Can you show the <layer-map> section of the lyt file that produced the above?

**Comment by maliberty** [comment]
Is the missing via found in either a LEF or DEF file that is input to klayout?

**Reply by 706697916** [reply]
below is the main part of my lyt file.
![1](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/54bd0d98-7b84-462c-afc9-916207386b9f)
![2](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/15917f3c-dff4-4b38-be4f-def034c3c087)
![3](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/assets/70256434/314254ab-881d-4614-a6a7-8ce1fd06de5d)

**Reply by 706697916** [reply]
@maliberty The missing via can be found in the tech.lef file or the 6_final.def file. The def file is the output of QRFS, so I think it is the input of Klayout. As for the lef file, I write it to the lyt file, it will report the same error of no matching gds cell.

**Reply by rovinski** [reply]
I think your syntax for the layers isn't correct. The layer name should come from the LEF file, not anywhere else. The LEF name is very likely `M1` instead of `M1.DRAWING`. Additionally, KLayout appends `.PIN` and `.LABEL` to the name when importing from DEF to differentiate DEF pins and add labels. So your layer map should likely look something like:
```
'M1 : 101/1';'M1.PIN : 101/1';'M1.LABEL : 101/6';'M2 : 102/1';'M2.PIN : 102/1';'M2.LABEL : 102/6';
...
```
etc

**Reply by 706697916** [reply]
@rovinski Thank you, it works! I also need to change the VIA name. Which lef file is used for the layer name, is it tech.lef?
